Applied mutation testing to improve the quality of the tests and the code for Java implementation.

omgm opened the pull request:
Applied [mutation testing](http://confreaks.tv/videos/mwrc2014-re-thinking-regression-testing) with Ortask Mutator 1.9. Improved the quality of the tests (and the code) by 13% (from a score of 84.6% to 95.3%).
- Explicitly made public fields 'public'.
- Corrected the comparison of the error value by introducing bounds for it (added a new error precision field). 
  -- Best programming practice: floats/doubles should never be compared like ints.
- Made the new error precision part of the public API.
- Added needed tests to document behavior with bad and edge-case parameters.

omgm committed:
floats/doubles should never be compared like ints

Applied [mutation testing](http://confreaks.tv/videos/mwrc2014-re-thinking-regression-testing) with Ortask Mutator 1.9. Improved the quality of the tests (and the code) by 13% (from a score of 84.6% to 95.3%).

- Explicitly made public fields 'public'.
- Corrected the comparison of the error value by introducing bounds for it (added a new error precision field). 
    -- Best programming practice: floats/doubles should never be compared like ints.
- Made the new error precision part of the public API.

omgm committed:
New tests for bad and edge-case parameters.

Applied [mutation testing](http://confreaks.tv/videos/mwrc2014-re-thinking-regression-testing) with Ortask Mutator 1.9. Improved the quality of the tests (and the code) by 13% (from a score of 84.6% to 95.3%).

- Added needed tests to document behavior with bad and edge-case parameters.

bergundy commented on java/src/main/java/me/everything/inbloom/BloomFilter.java:
```diff
@@ -24,11 +24,12 @@
-    int bits;
-    int bytes;
-    int hashes;
+    public int entries; 
```
If these are public, might as well make them final

dvirsky replied:
@bergundy they were package private before this PR. If the test is running in the same package, I suggest we keep them package-private, and final is just a nice addition.

bergundy replied:
let's go with protected

bergundy replied:
and final

bergundy replied:
@omgm please change the attributes to final protected

omgm replied:
Sorry guys, didn't know you were waiting on me. Changes have been done.

bergundy commented on java/src/main/java/me/everything/inbloom/BloomFilter.java:
```diff
@@ -24,11 +24,12 @@
+    public int bits;
+    public int bytes;
+    public int hashes;
+    public static double errorPrecision = 0.000000001;
```
Makes sense, @dvirsky ?

dvirsky replied:
yes but it means changing this in all 3 implementations as well

bergundy commented on java/src/test/java/me/everything/inbloom/BloomFilterTest.java:
```diff
@@ -9,6 +9,102 @@
+    {
+        try 
+        {
+            BloomFilter_mutated bf = new BloomFilter_mutated( -1, 0.01 );
```
Where does the BloomFilter_mutated class come from?
I don't understand

dvirsky replied:
I don't see `class BloomFilter_mutated` defined anywhere. Am I missing something?

omgm replied:
Oops! An artifact of mutation analysis. I'll fix it shortly.

omgm replied:
Done. Let me know if you see the changes. 
Good catch btw! This was my own mistake by not running the tests on the final copy. Lesson learned.

bergundy commented on java/src/test/java/me/everything/inbloom/BloomFilterTest.java:
```diff
@@ -9,6 +9,102 @@
 public class BloomFilterTest extends TestCase {
 
+    public void testCreateFilterWithBadParameters()
+    {
```
I understand that there's no convention for braces in the project, I think we should go with omitting the new line before opening braces.
I'll fix in the rest of the project.

dvirsky replied:
@bergundy oh there is - no newline before opening brace

omgm committed:
Cleaned up the new tests, adhered to convention

For mutation analysis, I create a new test to leave the original one intact. I forgot to clean up the new tests when I committed. This has been fixed.

omgm committed:
Made client-accessible fields 'protected final'

Made client-accessible fields 'protected final' as agreed.

bergundy commented:
@omgm merged, thanks!
